File: backend/src/datasource/dbdao/identity.rs
#![allow(dead_code)]
use super::dao::DBDao;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use sqlx::Row;

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct UserRecord {
    pub id: i64,
    pub username: String,
    pub email: String,
    pub password: Option<String>,
    pub role: Option<String>,
    pub enabled: bool,
    pub created_at: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct RoleRecord {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub created_at: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct PermissionRecord {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub created_at: Option<String>,
}

impl DBDao {
    /// Get user by ID
    pub async fn get_user(&self, id: i64) -> Result<UserRecord> {
        let row = sqlx::query(
            "SELECT id, username, email, password, role, enabled, created_at
             FROM users WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let user = UserRecord {
            id: row.get("id"),
            username: row.get("username"),
            email: row.get("email"),
            password: row.get("password"),
            role: row.get("role"),
            enabled: row.get("enabled"),
            created_at: row.get("created_at"),
        };

        Ok(user)
    }

    /// Search users with pagination
    pub async fn search_users(
        &self,
        username: Option<String>,
        _email: Option<String>,
        _role: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<UserRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, username, email, password, role, enabled, created_at
             FROM users
             WHERE ($1 IS NULL OR username = $1)
             OFFSET $2 LIMIT $3",
        )
        .bind(username)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let users = rows
            .into_iter()
            .map(|row| UserRecord {
                id: row.get("id"),
                username: row.get("username"),
                email: row.get("email"),
                password: row.get("password"),
                role: row.get("role"),
                enabled: row.get("enabled"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(users)
    }

    /// Add user
    pub async fn add_user(
        &self,
        username: &str,
        email: &str,
        password: Option<String>,
        role: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO users (username, email, password, role)
             VALUES ($1, $2, $3, $4)
             RETURNING id",
        )
        .bind(username)
        .bind(email)
        .bind(password)
        .bind(role)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update user
    pub async fn update_user(
        &self,
        id: i64,
        username: Option<String>,
        email: Option<String>,
        password: Option<String>,
        role: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE users
             SET username = COALESCE($1, username),
                 email = COALESCE($2, email),
                 password = COALESCE($3, password),
                 role = COALESCE($4, role)
             WHERE id = $5
             RETURNING id",
        )
        .bind(username)
        .bind(email)
        .bind(password)
        .bind(role)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Get role by ID
    pub async fn get_role(&self, id: i64) -> Result<RoleRecord> {
        let row = sqlx::query(
            "SELECT id, name, description, created_at
             FROM roles WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let role = RoleRecord {
            id: row.get("id"),
            name: row.get("name"),
            description: row.get("description"),
            created_at: row.get("created_at"),
        };

        Ok(role)
    }

    /// Search roles with pagination
    pub async fn search_roles(
        &self,
        name: Option<String>,
        _description: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<RoleRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, name, description, created_at
             FROM roles
             WHERE ($1 IS NULL OR name = $1)
             OFFSET $2 LIMIT $3",
        )
        .bind(name)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let roles = rows
            .into_iter()
            .map(|row| RoleRecord {
                id: row.get("id"),
                name: row.get("name"),
                description: row.get("description"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(roles)
    }

    /// Add role
    pub async fn add_role(&self, name: &str, description: Option<String>) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO roles (name, description)
             VALUES ($1, $2)
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update role
    pub async fn update_role(
        &self,
        id: i64,
        name: Option<String>,
        description: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE roles
             SET name = COALESCE($1, name),
                 description = COALESCE($2, description)
             WHERE id = $3
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Get permission by ID
    pub async fn get_permission(&self, id: i64) -> Result<PermissionRecord> {
        let row = sqlx::query(
            "SELECT id, name, description, created_at
             FROM permissions WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let permission = PermissionRecord {
            id: row.get("id"),
            name: row.get("name"),
            description: row.get("description"),
            created_at: row.get("created_at"),
        };

        Ok(permission)
    }

    /// Search permissions with pagination
    pub async fn search_permissions(
        &self,
        name: Option<String>,
        _description: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<PermissionRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, name, description, created_at
             FROM permissions
             WHERE ($1 IS NULL OR name = $1)
             OFFSET $2 LIMIT $3",
        )
        .bind(name)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let permissions = rows
            .into_iter()
            .map(|row| PermissionRecord {
                id: row.get("id"),
                name: row.get("name"),
                description: row.get("description"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(permissions)
    }

    /// Add permission
    pub async fn add_permission(&self, name: &str, description: Option<String>) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO permissions (name, description)
             VALUES ($1, $2)
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update permission
    pub async fn update_permission(
        &self,
        id: i64,
        name: Option<String>,
        description: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE permissions
             SET name = COALESCE($1, name),
                 description = COALESCE($2, description)
             WHERE id = $3
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }
}

File: backend/datasource/dbdao/identity.go
package dbdao

import (
	"context"
	"errors"
	"time"

	"gorm.io/gorm"
)

type IdentityUserModel struct {
	ID        int64     `gorm:"primarykey;column:id"`
	Username  string    `gorm:"column:username;type:varchar(100);not null;uniqueIndex"`
	Email     string    `gorm:"column:email;type:varchar(255);not null"`
	Password  string    `gorm:"column:password;type:varchar(255)"`
	Role      string    `gorm:"column:role;type:varchar(50)"`
	Enabled   bool      `gorm:"column:enabled;default:true"`
	CreatedAt time.Time `gorm:"column:created_at;autoCreateTime:milli"`
}

func (IdentityUserModel) TableName() string {
	return "identity_users"
}

type IdentityUser struct {
	ID        int64
	Username  string
	Email     string
	Password  string
	Role      string
	Enabled   bool
	CreatedAt time.Time
}

func (d *DB) GetIdentityUser(ctx context.Context, id int64) (*IdentityUser, error) {
	var model IdentityUserModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("user not found")
		}
		return nil, result.Error
	}
	return &IdentityUser{
		ID:        model.ID,
		Username:  model.Username,
		Email:     model.Email,
		Password:  model.Password,
		Role:      model.Role,
		Enabled:   model.Enabled,
		CreatedAt: model.CreatedAt,
	}, nil
}

func (d *DB) SearchIdentityUsers(ctx context.Context, username, email, role string, page, limit int64) ([]*IdentityUser, int64, error) {
	var models []IdentityUserModel
	q := d.DB().WithContext(ctx)
	if username != "" {
		q = q.Where("username LIKE ?", "%"+username+"%")
	}
	if email != "" {
		q = q.Where("email LIKE ?", "%"+email+"%")
	}
	if role != "" {
		q = q.Where("role = ?", role)
	}
	var total int64
	q.Model(&IdentityUserModel{}).Count(&total)
	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}
	users := make([]*IdentityUser, len(models))
	for i, m := range models {
		users[i] = &IdentityUser{
			ID:        m.ID,
			Username:  m.Username,
			Email:     m.Email,
			Password:  m.Password,
			Role:      m.Role,
			Enabled:   m.Enabled,
			CreatedAt: m.CreatedAt,
		}
	}
	return users, total, nil
}

func (d *DB) CreateIdentityUser(ctx context.Context, user *IdentityUser) (*IdentityUser, error) {
	model := &IdentityUserModel{
		Username: user.Username,
		Email:    user.Email,
		Password: user.Password,
		Role:     user.Role,
		Enabled:  user.Enabled,
	}
	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}
	return &IdentityUser{
		ID:        model.ID,
		Username:  model.Username,
		Email:     model.Email,
		Password:  model.Password,
		Role:      model.Role,
		Enabled:   model.Enabled,
		CreatedAt: model.CreatedAt,
	}, nil
}

func (d *DB) UpdateIdentityUser(ctx context.Context, user *IdentityUser) (*IdentityUser, error) {
	updates := map[string]interface{}{}
	if user.Username != "" {
		updates["username"] = user.Username
	}
	if user.Email != "" {
		updates["email"] = user.Email
	}
	if user.Password != "" {
		updates["password"] = user.Password
	}
	if user.Role != "" {
		updates["role"] = user.Role
	}
	updates["enabled"] = user.Enabled
	result := d.DB().WithContext(ctx).Model(&IdentityUserModel{}).Where("id = ?", user.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}
	if result.RowsAffected == 0 {
		return nil, errors.New("user not found")
	}
	return d.GetIdentityUser(ctx, user.ID)
}

type RoleModel struct {
	ID          int64     `gorm:"primarykey;column:id"`
	Name        string    `gorm:"column:name;type:varchar(100);not null;uniqueIndex"`
	Description string    `gorm:"column:description;type:varchar(500)"`
	CreatedAt   time.Time `gorm:"column:created_at;autoCreateTime:milli"`
}

func (RoleModel) TableName() string {
	return "identity_roles"
}

type Role struct {
	ID          int64
	Name        string
	Description string
	CreatedAt   time.Time
}

func (d *DB) GetRole(ctx context.Context, id int64) (*Role, error) {
	var model RoleModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("role not found")
		}
		return nil, result.Error
	}
	return &Role{
		ID:          model.ID,
		Name:        model.Name,
		Description: model.Description,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) SearchRoles(ctx context.Context, name, description string, page, limit int64) ([]*Role, int64, error) {
	var models []RoleModel
	q := d.DB().WithContext(ctx)
	if name != "" {
		q = q.Where("name LIKE ?", "%"+name+"%")
	}
	var total int64
	q.Model(&RoleModel{}).Count(&total)
	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}
	roles := make([]*Role, len(models))
	for i, m := range models {
		roles[i] = &Role{
			ID:          m.ID,
			Name:        m.Name,
			Description: m.Description,
			CreatedAt:   m.CreatedAt,
		}
	}
	return roles, total, nil
}

func (d *DB) CreateRole(ctx context.Context, role *Role) (*Role, error) {
	model := &RoleModel{
		Name:        role.Name,
		Description: role.Description,
	}
	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}
	return &Role{
		ID:          model.ID,
		Name:        model.Name,
		Description: model.Description,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) UpdateRole(ctx context.Context, role *Role) (*Role, error) {
	updates := map[string]interface{}{}
	if role.Name != "" {
		updates["name"] = role.Name
	}
	if role.Description != "" {
		updates["description"] = role.Description
	}
	result := d.DB().WithContext(ctx).Model(&RoleModel{}).Where("id = ?", role.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}
	if result.RowsAffected == 0 {
		return nil, errors.New("role not found")
	}
	return d.GetRole(ctx, role.ID)
}

type PermissionModel struct {
	ID          int64     `gorm:"primarykey;column:id"`
	Name        string    `gorm:"column:name;type:varchar(100);not null;uniqueIndex"`
	Description string    `gorm:"column:description;type:varchar(500)"`
	CreatedAt   time.Time `gorm:"column:created_at;autoCreateTime:milli"`
}

func (PermissionModel) TableName() string {
	return "identity_permissions"
}

type Permission struct {
	ID          int64
	Name        string
	Description string
	CreatedAt   time.Time
}

func (d *DB) GetPermission(ctx context.Context, id int64) (*Permission, error) {
	var model PermissionModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("permission not found")
		}
		return nil, result.Error
	}
	return &Permission{
		ID:          model.ID,
		Name:        model.Name,
		Description: model.Description,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) SearchPermissions(ctx context.Context, name, description string, page, limit int64) ([]*Permission, int64, error) {
	var models []PermissionModel
	q := d.DB().WithContext(ctx)
	if name != "" {
		q = q.Where("name LIKE ?", "%"+name+"%")
	}
	var total int64
	q.Model(&PermissionModel{}).Count(&total)
	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}
	perms := make([]*Permission, len(models))
	for i, m := range models {
		perms[i] = &Permission{
			ID:          m.ID,
			Name:        m.Name,
			Description: m.Description,
			CreatedAt:   m.CreatedAt,
		}
	}
	return perms, total, nil
}

func (d *DB) CreatePermission(ctx context.Context, perm *Permission) (*Permission, error) {
	model := &PermissionModel{
		Name:        perm.Name,
		Description: perm.Description,
	}
	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}
	return &Permission{
		ID:          model.ID,
		Name:        model.Name,
		Description: model.Description,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) UpdatePermission(ctx context.Context, perm *Permission) (*Permission, error) {
	updates := map[string]interface{}{}
	if perm.Name != "" {
		updates["name"] = perm.Name
	}
	if perm.Description != "" {
		updates["description"] = perm.Description
	}
	result := d.DB().WithContext(ctx).Model(&PermissionModel{}).Where("id = ?", perm.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}
	if result.RowsAffected == 0 {
		return nil, errors.New("permission not found")
	}
	return d.GetPermission(ctx, perm.ID)
}
